WordPress/gutenberg
# User autocomplete - use debounced search request

## Description
Looking into user autocomplete (typing `@` and then starting to type a username, see https://github.com/WordPress/gutenberg/pull/2896), specifically the code for `userAutocompleter`: currently, it attempts to load all users with a single fetch via `new wp.api.collections.Users() ).fetch().then( ( users ) => {...`.  While this works for testing with a few users, it is not scalable. By default the REST API users endpoint returns 10 users and the default maximum for REST requests is 100 records. Some sites might have hundreds or even thousands of users, so the filtering applied as you type won't work.

This should instead use a debounced search. Every time the user types and pauses briefly, fire off a request to the REST API with the typed search string, this will work: `new wp.api.collections.Users() ).fetch( { data: { search: searchString } } ).then( ( users ) => {...`. 

Addresses https://github.com/WordPress/gutenberg/issues/2793#issuecomment-356822668

## How Has This Been Tested/How to test
* Load a post and type `@` and start typing a username. 
* Check the network console, you will see requests being sent with a search string
* Note the requests are sent at most every 250ms which should feel immediate while preventing over requesting

## Screenshots (jpeg or gifs if applicable):
![](https://cl.ly/0Y342y360p2B/Image%202018-01-11%20at%208.55.44%20AM.png)

## Types of changes
* debounce search callback is autocompleters/index.js
* in search callback, in autocompleter popover is open and selected autocompleter exposes a `setSearch` function, calls this function with the current search string on every (debounced)  input event

## Checklist:
